rfc/rfc017-distributed-network-grpc-v2.md

```diff
@@ -45,4 +54,154 @@ As specified by RFC015, the node MUST authenticate the peer's node DID as follow
+     * Lamport Clock start value.
+     * Lamport Clock end value.
+
+4. When receiving a request for transactions, Bob responds with a message including the requested transactions (§4.2.4 and §4.2.5).
+
+5. After adding Bob's transactions to the DAG \(making sure its cryptographic signature is valid\), query the payload if it's missing.
+
+A node MUST make sure to only add transactions of which all previous transactions are present.
+
+The sections below specify the details of the protocol operation and maps it to the gRPC messages. See the section "Protobuf Definition" for a full specification of the protobuf/gRPC contract.
+
+### 4.2.1 Broadcasting
+
+The local node's DAG state MUST be broadcast at an interval using the `AdvertState` message, by default every 2 seconds. The interval MAY be adjusted by the node operator but MUST conform to the limits \(min/max interval\) defined by the network. It is advised to keep it relatively short because it directly influences the speed by which new transactions are retrieved.
+
+The `AdvertState` message contains a `XOR` value, an `LC` value and a `token`. The `token` is scoped to a connection and is required in the response from the peer.
+The `LC` value MUST equals the highest Lamport Clock value of all transaction references included in the `XOR` calculation.
+If no transactions are present, an all-zero `XOR` and `LC` of 0 is sent.
+A node is free to choose how the `token` is created as long as it's able to satisfy the requirements of §4.2.3.
+It MUST be valid for at least 10 seconds.
+
+### 4.2.2 IBLT response
+
+When a peer receives an `AdvertState` message and the `XOR` differs, it SHOULD respond with a `TransactionSet` message.
+The upper bound of the response IBLT MUST be the equal to `PSIZE*ceil(LC/PSIZE)` where `LC` is the value from the `AdvertState` message and `ceil` is a function that rounds a double up to the nearest integer.
+If `LC` equals zero, `PSIZE` is used.
+This upper bound is exclusive, the lower bound is inclusive and always 0.
+Next to the IBLT data, the `TransactionSet` message MUST contain the original `LC` value as `LC_req`, a new `LC` value indicating the highest LC value from the peer and the `token` value from the `AdvertState` message.
+
+### 4.2.3 Transaction List Query
+
+For every `TransactionSet` message a node receives, it MUST check the `token` value in the message and compare it to its own administration.
+If the `token` doesn't exist or if the `LC_req` value doesn't match, the node MUST ignore the message.
+
+The IBLT in the `TransactionSet` message contains every transaction of the peer in the LC range of `0-LC_req`. The node MUST lookup/compute the IBLT for this range and subtract it from the IBLT of the peer. Deconstruction of the resulting IBLT will list the transaction refs the peer has and the local node misses. 
+
+These transactions MUST be queried by using the `TransactionListQuery` message. It contains all missing transaction references.
```

Comment:
if we don't allow adding TXs with missing prevs to the DAG, you'll almost never query "random" TXs (only in case of a branch), so wouldn't range queries suffice?

Reply:
If someone creates a branch at LC=0 while your LC is at 391867540913, then you have a big problem if your only tool is the range query. It could also have created a TX with prevs=[0, 391867540913]. With the IBLT you can exactly query those 2 TXs, with ranges not so much....